src/main/java/org/dbos/apiary/xa/MySQLXAConnection.java

```diff
@@ -0,0 +1,198 @@
+package org.dbos.apiary.xa;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mysql.cj.jdbc.MysqlXADataSource;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+public class MySQLXAConnection implements XADBConnection {
+    private static final Logger logger = LoggerFactory.getLogger(MySQLXAConnection.class);
+
+    private final ThreadLocal<MysqlXADataSource> ds;
+    private final ThreadLocal<javax.sql.XAConnection> xaconnection;
+    private final ThreadLocal<Connection> connection;
+
+    public MySQLXAConnection(String hostname, Integer port, String databaseName, String databaseUsername, String databasePassword) throws SQLException {
+        this.ds = ThreadLocal.withInitial(() -> {
```

Comment:
I think DataSource might be thread-safe so you may not need a thread-local variable here.

Reply:
Fixed